### PR TITLE
Allow other plugins to create renderer resources

### PR DESCRIPTION
**Pull request by awtterpip**
# Objective

- Allow other plugins to create the renderer resources. An example of where this would be required is my  [OpenXR plugin](https://github.com/awtterpip/bevy_openxr)

## Solution

- Changed the bevy RenderPlugin to optionally take precreated render resources instead of a configuration.

## Migration Guide

The `RenderPlugin` now takes a `RenderSettings` enum instead of `WgpuSettings`. `RenderSettings::default()` returns `RenderSettings::Automatic(WgpuSettings::default())`.

```rust
// before
RenderPlugin {
    wgpu_settings: WgpuSettings {
    ...
    },
}

// now
RenderPlugin {
    render_settings: RenderSettings::Automatic(WgpuSettings {
    ...
    }),
}
```

### CODE REVIEW

File: crates/bevy_render/src/lib.rs
Comment (JMS55) [inline crates/bevy_render/src/lib.rs:193]: Struct needs documentation.

File: crates/bevy_render/src/lib.rs
Comment (JMS55) [inline crates/bevy_render/src/lib.rs:220]: Can we extract this chunk of code into a function for initializing the render resources?

File: crates/bevy_render/src/lib.rs
Comment (JMS55) [inline crates/bevy_render/src/lib.rs:194]: I know this is unrelated to this PR, but rather than a Mutex, I feel like this should be a OnceLock. Thoughts?

**Review by JMS55 — CHANGES_REQUESTED** [review]
I'm for these changes overall. It'll also make DLSS nicer when I get around to implementing that again :)

I do have some small code refactor suggestions that would be nice, if you don't mind doing them while you're here.

File: crates/bevy_render/src/lib.rs
Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:194]: I have added documentation. If this documentation is not good enough, or you have better doc comments in mind, please let me know.

File: crates/bevy_render/src/lib.rs
Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:221]: I have extracted this piece of code into a different function. I was debating between having backends be a variable for the render resource initialization, or simply unwrapping the backends field of settings since we can be sure that it will be Some. If anymore changes need to be made please let me know.

File: crates/bevy_render/src/lib.rs
Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:195]: While I understand how a OnceLock would make more sense here, to extract the renderer resources outside of the Arc reference, we need a struct with interior mutability since wgpu::Instance does not implement Clone.

**Review by JMS55 — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

File: crates/bevy_render/src/lib.rs
Comment (paul-hansen) [inline crates/bevy_render/src/lib.rs:194]: What would modifying this resource at runtime do? 

If nothing or it would break things, would it make more sense to add these as an optional field to the RenderApp that RenderApp would use to init this resource instead? Seems like that might be a better way to go for one time configuration?

Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:194]: it would be very unlikely to break anything. the only time that resource ever matters is during the plugin build stage. If a plugin replaces that resource after the render crate has created it, nothing bad will happen except it will use the plugin's render resources. i just recommend initializing them before because it will let the render crate skip wgpu initialization.

Comment (paul-hansen) [reply crates/bevy_render/src/lib.rs:194]: Cool, so if they only do anything at initialization I'd probably recommend not exposing the resource (which could have been modified at any time) and instead allow them to be provided as optional fields when initializing the plugin. 

See the asset plugin as an example of this https://docs.rs/bevy/latest/bevy/asset/struct.AssetPlugin.html you can provide optional watch for changes settings.

For why this is better, 
- it makes it clearer to users how to use it just by looking at the API because it isn't possible to set it at the wrong time.
- prevents multiple third party crates from trying to all create and set it without the user realizing.
- makes it more obvious to the user that this is something that can be configured with the render app.

Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:194]: mhm, how do you think i should implement it? should i add a `Option<FutureRenderResources>` field to the render plugin, or should i make that struct private again and have people pass the render resources to the render plugin? i think i should go with the first option, since the render resources aren't actually initialized until the plugin build phase.

Comment (paul-hansen) [reply crates/bevy_render/src/lib.rs:194]: Hmm not sure, I was leaning towards it just being a field that takes an option with a tuple, or multiple fields if it makes sense to only provide some of them sometimes. I haven't really dug into the way they get initialized or anything so idk if multiple fields would make sense. I would lean towards not exposing a resource so people don't think they can do something with that resource. Usually when I see a public resource I assume I can do something with it at runtime. I'll leave that up to you though, I try not to nit pick too much.

Also, this was just a thought I had while reviewing and I'm not really anyone particularly important here lol. Feel free to disagree after considering it.
Looking forward to more VR Bevy stuff so I don't mean to hold things up!

Comment (alice-i-cecile) [reply crates/bevy_render/src/lib.rs:194]: An optional field on RenderPlugin would be nice. Definitely no public resource if it can't be set at runtime.

Similarly, it might be helpful to allow specifying only some of these settings if partial initalization is meaningful.

File: crates/bevy_render/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/lib.rs:379]: We can't validate this safety invariant purely from the function signature: you could get a `&mut App` on any thread. As a result, this whole function needs to be unsafe, with the safety requirements plumbed up.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I think we can and should make the API better, by swapping to a "field on RenderPlugin" strategy.

More importantly though, the extracted function needs to be `unsafe`.

File: crates/bevy_render/src/lib.rs
Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:194]: right now the problem i'm having though is that usually the render resources aren't actually initialized until the build phase of the plugin, however, having them as an optional field on the RenderPlugin would require the render resources to be already initialized when the plugin is made

Comment (paul-hansen) [reply crates/bevy_render/src/lib.rs:194]: I was thinking that for a third party crate to set these you would have users disable Bevy's render plugin from DefaultPlugins and then add the third party's render plugin which would simply init the render resources and then add Bevy's render plugin back with those resources in the third party plugin's build step.

The third party crate could also provide their own DefaultPlugins that does something similar if they want to simplify what the user has to do a bit.

Would that work?

Comment (awtterpip) [reply crates/bevy_render/src/lib.rs:194]: that sounds like it would work perfectly,

**Comment by awtterpip** [comment]
i believe i have implemented a better way to accomplish this

File: crates/bevy_render/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/lib.rs:392]:
```suggestion
/// SAFETY: this function must be called from the main thread.
unsafe fn initialize_render_app(app: &mut App) {
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/settings.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/settings.rs:96]: It might be nice to link back to `RenderPlugin` here, to better explain how this could be used.

**Comment by alice-i-cecile** [comment]
I like the new initialization pattern a lot! Once there's a simple migration guide plus a SAFETY comment, this LGTM.

**Comment by awtterpip** [comment]
where should i write the migration guide?

**Comment by paul-hansen** [comment]
It goes in the PR description, here's an example from one of mine: https://github.com/bevyengine/bevy/pull/9516

The migration guide gets created from a script that scrapes these from the PRs so make sure the header is exact. E.g.
```
## Migration Guide

Instructions for migrating code that will break
```

File: crates/bevy_render/src/settings.rs
Comment (mockersf) [inline crates/bevy_render/src/settings.rs:58]: this should be kept

**Comment by superdump** [comment]
I want to carefully review this before it is merged. What is the reason for needing to provide the ‘future render resources’ from outside bevy? What configurability is missing?

**Comment by awtterpip** [comment]
@superdump so the solution ended up actually changing. originally, i made `FutureRendererResources` a public struct, and i made the RenderPlugin skip creating the render resources if this already existed, this would allow other plugins to create these resources themselves. however, the solution has changed to making the RenderPlugin be able to take precreated render resources to prevent the resources being able to be edited at runtime. An example of where this would be needed is my openxr plugin for bevy. openxr has a very specific process to initalize render resources, requiring me to make these resources myself and then give them to the bevy RenderPlugin

**Comment by JMS55** [comment]
> What is the reason for needing to provide the ‘future render resources’ from outside bevy? What configurability is missing?

It's mainly for importing rendering resources from outside of bevy. For instance, with XR, you have a specific vulkan resource, that you then map to a wgpu resource in a specific manner. For DLSS, I needed to use raw vulkan to load some extra nvidia-dlss-only extensions (so, stuff wgpu would never get) as part of device creation, and then create a wgpu device wrapping it.

It's the kind of stuff bevy_render is never going to handle itself, for very specific platforms with custom rendering devices.

**Comment by awtterpip** [comment]
i believe that is a satisfactory migration guide. if it needs improving please let me know

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_render/src/settings.rs
Comment (awtterpip) [reply crates/bevy_render/src/settings.rs]: fixed

**Review by MalekiRe — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Waiting on final review from `@superdump` :) It's in the 0.12 milestone, so it *will* be considered.

File: crates/bevy_render/src/settings.rs
Comment (mockersf) [inline crates/bevy_render/src/settings.rs:108]: the mutex here is to allow interior mutability, which breaks the assumption that setting up a plugin doesn't mutate it

Comment (TheButlah) [reply crates/bevy_render/src/settings.rs:108]: from what I could tell, all methods on `wgpu::Instance` are on a shared ref, and the type is `Send` and `Sync`. Is there a reason that cloning a `Arc<wgpu::Instance>` cannot be used instead taking from a `Mutex<Instance>`?

Comment (awtterpip) [reply crates/bevy_render/src/settings.rs]: @mockersf my most recent commit should have fixed this. i implemented what @TheButlah recommended

**Comment by awtterpip** [comment]
i had to rebuild my pr since i was having some merge conflicts and forgot to make a new branch when i started making this pr, however, github won't let me reopen this PR without the original branch, so i had to create a duplicate of this pr here: #9925

**Pull request closed without merging**
